platform/m6/sub/101a/boot.c

```diff
@@ -0,0 +1,251 @@
+			"    str       r2, [r1, #0]\n"
+			"    bne.n     pcr1\n"
+			"    dsb       sy\n"
+			"    bl        enable_branch_prediction\n"
+			"    bx        r3\n"
+			"disable_branch_prediction:\n"
+			"    mrc       p15, 0, r1, c1, c0, 0\n"
+			"    bic.w     r1, r1, #0x800\n"
+			"    mcr       p15, 0, r1, c1, c0, 0\n"
+			"    dsb       sy\n"
+			"    isb       sy\n"
+			"    bx        lr\n"
+			"enable_branch_prediction:\n"
+			"    mcr       p15, 0, r1, c7, c5, 6\n"
+			"    dsb       sy\n"
+			"    mrc       p15, 0, r1, c1, c0, 0\n"
+			"    orr.w     r1, r1, #0x800\n"
+			"    mcr       p15, 0, r1, c1, c0, 0\n"
+			"    dsb       sy\n"
+			"    isb       sy\n"
+			"    bx        lr\n"
+	);
+}
+
+void __attribute__((naked,noinline)) sub_e00200f8_my() {
+	//  capdis -f=chdk -jfw -stubs -s=0xe00200f9 -c=2 PRIMARY.BIN 0xe0000000
+	asm volatile (
+			"    push    {r4, r5, r6, lr}\n"
+			"    ldr     r0, =0x002b413f\n" // heap start, modify here
+			"    ldr     pc, =0xe00200fd\n"// Continue in firmware
+			".ltorg\n"
+	);
+}
+
+/**
+ * @see main startup
+ */
+void __attribute__((naked,noinline)) boot() {
+	// capdis -f=chdk -stubs -s=0xe0020011 -c=6 PRIMARY.BIN 0xe0000000
+	asm volatile (
+			"    ldr     r0, =0xe0020200\n"
+			"    mcr     p15, #0, r0, c12, c0, #0\n"
+			"    isb     sy\n"
+			"    movw    r0, #0x2000\n"
+			"    movt    r0, #0\n"
+			"    mov     sp, r0\n"
+	);
+
+	// handle 2 cores
+	asm volatile (
+			"    mrc     p15, #0, r0, c0, c0, #5\n"
+			"    ands    r0, #0xf\n"
+			"    beq     loc_boot\n"
+			"    adr     r0, loc_boot\n"
+			"    orr     r0, #1\n"
+			"    bl      park_current_core_then_continue\n"
+			"loc_boot:\n"
+			"    mrc     p15, #0, r0, c0, c0, #5\n"
+			"    ands    r0, #0xf\n"
+			"    bl      setup_mmu_for_chdk\n"
+	);
+
+	// capdis -f=chdk -stubs -s=0xe0020025 -c=59 PRIMARY.BIN 0xe0000000
+	asm volatile (
+			"    mrc     p15, #0, r5, c0, c0, #5\n"
+			"    ands    r0, r5, #0xf\n"
+			"    bne     loc_e0020032\n"
+			"    b.w     loc_e002003c\n"
+			"loc_e0020032:\n"
+			"    b.w     sub_e00200f8_my\n"
+			"loc_e002003c:\n"
+			"    ldr     r0, =0xe115d544\n"
+			"    ldr     r1, =0x00008000\n"
+			"    ldr     r3, =0x00068524\n"
+			"loc_e0020042:\n"
+			"    cmp     r1, r3\n"
+			"    itt     lo\n"
+			"    ldrlo   r2, [r0], #4\n"
+			"    strlo   r2, [r1], #4\n"
+			"    blo     loc_e0020042\n"
+			"    ldr     r1, =0x002b4138\n"
+			"    mov.w   r2, #0\n"
+			"loc_e0020056:\n"
+			"    cmp     r3, r1\n"
+			"    it      lo\n"
+			"    strlo   r2, [r3], #4\n"
+			"    blo     loc_e0020056\n"
+			"    ldr     r0, =0xe11bda68\n"
+			"    ldr     r1, =0x01900000\n"
+			"    ldr     r3, =0x01901444\n"
+			"loc_e0020066:\n"
+			"    cmp     r1, r3\n"
+			"    itt     lo\n"
+			"    ldrlo   r2, [r0], #4\n"
+			"    strlo   r2, [r1], #4\n"
+			"    blo     loc_e0020066\n"
+			"    ldr     r0, =0x01900000\n"
+			"    ldr     r1, =0x00001444\n"
+			"    bl      _dcache_clean_by_mva\n"
```

Comment:
if the stub method is used then it's a stub address (real address + 1) from the stubs_entry_2.S ?  ...does it matter ?

Reply:
When the PC register is loaded, bit0 of the address specifies the instruction set. Most of the firmware is Thumb code, so most stubs have their bit0 set. Stub addresses appearing in stubs_entry files are loaded in PC (the macros used there are defined in include/stubs_asm.h), so bit0 is important.

Reply:
so how I've done it; is it ok? instead of using sub_xxxxxx from the stubs_auto.S

Reply:
Yes. Capdis can do this for you if the stub is present in one of the stubs_ files and you call it with the -stubs switch.